TASK: New configuration syntax for request patterns & firewall filters

bwaidelich opened the pull request:
The configuration format for authentication request patterns is
not very flexible and inconsistent with other settings.

With this change the syntax is changed from:
```yaml
  'SomeAuthenticationProvider':
    # ...
    requestPatterns:
      'controllerObjectName': 'Some\Package\AdministrationArea\.*'
```

To a more generic syntax:

```yaml
  'SomeAuthenticationProvider':
    # ...
    requestPatterns:
      'Some.Package:AdministrationArea':
        pattern: 'ControllerObjectName'
        patternOptions:
          'controllerObjectNamePattern': 'Some\Package\AdministrationArea\.*'
```

The new syntax seems more cumbersome but it allows for advanced request
pattern options. Besides the new level allows to attach *multiple*
patterns of one type to a provider.
This also allows to *unset* previously configured patterns:

```yaml
  'SomeAuthenticationProvider':
    requestPatterns:
      'Some.Package:AdministrationArea': ~
```

The configuration syntax for request patterns for the firewall has
been adjusted accordingly::

```yaml
    TYPO3:
      Flow:
        security:
          firewall:
            filters:
              'Some.Package:BlockedUris':
                pattern: 'Uri'
                patternOptions:
                  'uriPattern': '\/some\/uri\/.*'
                interceptor: AccessDeny
```

This is not a breaking change as the old syntax is still supported.

But as the new syntax is recommended, this comes with a corresponding
core migration that you can execute as follows:

```
  /flow core:migrate --version 20151113161300 --verbose
```

FLOW-412 #close

aertmann commented:
excellent

bwaidelich commented on TYPO3.Flow/Migrations/Code/Version20151113161300.php:
```diff
@@ -0,0 +1,161 @@
+/**
+ * Adjust "Settings.yaml" to new "requestPattern" and "firewall" syntax (see FLOW-412)
+ */
+class Version20151113161300 extends AbstractMigration
```
This will adjust request patterns like this:
![image](https://cloud.githubusercontent.com/assets/307571/11151602/4fcb5d06-8a2e-11e5-9f3f-ba934e1339f7.png)
and firewall filters like:
![image](https://cloud.githubusercontent.com/assets/307571/11151608/594a171e-8a2e-11e5-98cd-495c75410694.png)


I'm not perfectly sure whether using the classname for the index is a good idea!?

skurfuerst replied:
Hey Bastian,

this feels rather strange for me; do I get it right that this is only done by the migration?

I'd rather have pattern identifiers like: "My.Package:Uri" or "My.Package:Ip" -- so it is not confused with Class names.

All the best,
Sebastian

bwaidelich replied:
Right, this is done by the migration and only for 3rd party packages that used either of the patterns or filters..

Good point, I wasn't sure about this one..
But unfortunately it's a bit tricky.

So for the filters:
Previously we used the array syntax (rather than a dictionary) leading to numeric keys.
So if someone specified one filter for example it would override the first (and only) defined filter of the Flow package, being the `CsrfProtection` (!).

I guess it makes sense to change it to `Package.Key:<Filter>`, so `TYPO3.Flow:CsrfProtection` in this case.
I doubt that a lot of people used this while relying on the fact that a filter overrides the filters with the same index..


For the request patterns it's a bit different:

It's quite likely that a 3rd party package *overrides* the pattern(options) for an existing auth provider.. In this case if we change the name to something other than the pattern classname we risk that the final configuration is sth like:

```
requestPatterns:
  'TYPO3.Neos:ControllerObjectName':
    pattern: 'ControllerObjectName'
    patternOptions:
      controllerObjectNamePattern: 'TYPO3\Neos\Controller\.*|TYPO3\Neos\Service\.*|TYPO3\Media\Controller\.*'
  'My.Package:ControllerObjectName':
    pattern: 'ControllerObjectName'
    patternOptions:
      controllerObjectNamePattern: 'Overridden...'

ending in a *merge* rather than the original *override*

skurfuerst commented on TYPO3.Flow/Classes/TYPO3/Flow/Security/Authentication/AuthenticationProviderManager.php:
```diff
@@ -318,11 +319,22 @@ protected function buildProvidersAndTokensFromConfiguration(array $providerConfi
-                foreach ($providerConfiguration['requestPatterns'] as $patternType => $patternConfiguration) {
+                foreach ($providerConfiguration['requestPatterns'] as $patternName => $patternConfiguration) {
+                    if (is_string($patternConfiguration)) {
+                        $patternType = $patternName;
```
I'd love a small comment explaining which parts of the code are needed for backwards compatibility; i.e. this one ;)

bwaidelich replied:
yes, good point

skurfuerst commented on TYPO3.Flow/Classes/TYPO3/Flow/Security/Authentication/AuthenticationProviderManager.php:
```diff
@@ -318,11 +319,22 @@ protected function buildProvidersAndTokensFromConfiguration(array $providerConfi
+                    if (!$requestPattern instanceof RequestPatternInterface) {
+                        throw new Exception\InvalidRequestPatternException(sprintf('Invalid request pattern configuration in setting "TYPO3:Flow:security:authentication:providers:%s": Class "%s" does not implement RequestPatternInterface', $providerName, $patternClassName), 1446222774);
+                    }
+                    if (is_string($patternConfiguration) && is_callable([$requestPattern, 'setPattern'])) {
```
... and this one

skurfuerst commented on TYPO3.Flow/Classes/TYPO3/Flow/Security/Authorization/FilterFirewall.php:
```diff
@@ -106,12 +106,18 @@ public function blockIllegalRequests(\TYPO3\Flow\Mvc\ActionRequest $request)
-            $requestPattern = $this->objectManager->get($this->requestPatternResolver->resolveRequestPatternClass($singleFilterSettings['patternType']));
-            $requestPattern->setPattern($singleFilterSettings['patternValue']);
+            $requestPattern = $this->objectManager->get($patternClassName, $patternOptions);
+            if (isset($singleFilterSettings['patternValue']) && is_callable([$requestPattern, 'setPattern'])) {
```
same here, add a comment which part is needed for backwards compat :)

skurfuerst commented on TYPO3.Flow/Classes/TYPO3/Flow/Security/RequestPattern/ControllerObjectName.php:
```diff
@@ -11,27 +11,30 @@
 
     /**
-     * Returns the set pattern
+     * Expects options in the form array('controllerObjectNamePatters' => '<regularExpression>')
```
typo: controllerObjectNamePatterNs

bwaidelich replied:
eagle eye kurfürst, awesome ;)

skurfuerst commented on TYPO3.Flow/Configuration/Settings.yaml:
```diff
@@ -380,21 +380,24 @@ TYPO3:
+          'TYPO3.Flow:CsrfProtection':
+            pattern: 'CsrfProtection'
+            patternOptions:
+              'foo': 'bar'
```
the foo:bar option should be removed, right? :)

bwaidelich replied:
Uhm ye

skurfuerst commented on TYPO3.Flow/Tests/Unit/Security/Authorization/FilterFirewallTest.php:
```diff
@@ -97,11 +97,99 @@ public function configuredFiltersAreCreatedCorrectly()
+
     /**
-     * @test
+     * __test
```
commented out test?

skurfuerst commented on TYPO3.Flow/Tests/Unit/Security/Authorization/FilterFirewallTest.php:
```diff
@@ -121,7 +209,7 @@ public function allConfiguredFiltersAreCalled()
 
     /**
-     * @test
+     * __test
```
commented out test?

bwaidelich replied:
ouch

skurfuerst commented:
Hey Bastian,

From my side, *really awesome* change -- and I think this will greatly improve the way we can use this.

Some comments:

* See inline comments :) Code quality looks quite good!
* Documentation updates are missing
* Mark the change from TASK to FEATURE

Furthermore, a question: Is it now possible to *extend* the RequestPattern? So far, I always needed the following line if wanting to write a controller which interacts with the Neos Backend:

```
TYPO3:
  Flow:
    security:
      authentication:
        providers:
          Typo3BackendProvider:
            requestPatterns:
              # !!! COPY+PASTE line from Neos and extend it by another OR.
              controllerObjectName: 'TYPO3\Neos\Controller\.*|TYPO3\Neos\Service\.*|TYPO3\Media\Controller\.*|Sandstorm\Newsletter\Controller\(NewsletterSending|ReceiverGroupModule)Controller.*'
```

Would I still need to do this, or can this be handled by another RequestPattern?

So my main question boils down to: Are RequestPatterns inside Auth providers ANDed (bad!) or ORed (good!) together?

If they are ANDed together, we should probably make the options for ControllerObjectName an array and OR it inside.

... the above is probably something for the Docs! (Neos :) )

All the best,
Sebastian

bwaidelich commented:
@skurfuerst Thanks a lot for your valuable feedback!

I addressed most comments with 093eab4a1aaf0bef3f0548908dc29f082dd3593f

And you're right. It seems request patterns are actually "ANDed":
https://github.com/neos/flow-development-collection/blob/master/TYPO3.Flow/Classes/TYPO3/Flow/Security/Context.php#L678-L685

I'd suggest to change that code to AND only by pattern type so that:

```yaml
  'Typo3BackendProvider':
    # ...
    requestPatterns:
      'My.Package:Foo':
        pattern: 'ControllerObjectName'
        patternOptions:
          'controllerObjectNamePattern': 'Some\Pattern'
```
will actually **add** the controller to the pattern and
```yaml
  'Typo3BackendProvider':
    # ...
    requestPatterns:
      'TYPO3.Neos:NeosControllers':
        patternOptions:
          'controllerObjectNamePattern': 'Some\Pattern'
```
will **override** it (given, it is called `TYPO3.Neos:NeosControllers` in the neos package)..
At least that was my original idea and you made me realize I didn't even test this case ;)

dfeyer commented:
I'm not sure if it's a valid use case, but the following configuration will allow to disable some controllerObjectName patterns from other package. I think about a package that replace some controller, per example.

```yaml
          Typo3BackendProvider:
            requestPatterns:
              'TYPO3.Neos:NeosControllers':
                patternOptions:
                  controllerObjectName:
                    'Neos\Media\Browser\Controller\Backend\.*': TRUE
```

kitsunet commented:
I like the direction this is going to, how can we proceed? Could someone fix the style issue?

RafaelKa commented:
Hey, can somebody understand what StyleCI exactly want?
@bwaidelich maybe double empty-line [ControllerObjectName.php](https://github.com/bwaidelich/flow-development-collection/blob/requestPattern-options/TYPO3.Flow/Classes/TYPO3/Flow/Security/RequestPattern/ControllerObjectName.php#L17-L18) and [](https://github.com/bwaidelich/flow-development-collection/blob/requestPattern-options/TYPO3.Flow/Classes/TYPO3/Flow/Security/RequestPattern/Uri.php#L17-L18)

kitsunet commented:
Yep, it's the empty line that needs to be taken out I guess

kitsunet commented:
mmm, @bwaidelich got no push rights to your fork :( You need to fix the stuff

kitsunet commented:
:+1:

bwaidelich commented:
@kitsunet weird, I added you as collaborator so you should have push rights (And didn't you do that with c3992c5 ?)
Unfortunately there's more to fix than just the styling issues.. I hope I'll find the time to tackle this at the sprint next week...

kitsunet commented:
ok cool, yeah this one thing was fixed :)

kitsunet commented:
@bwaidelich would you update this? :)

dfeyer commented:
@bwaidelich Can you rebase this one ?

bwaidelich commented:
@kitsunet @dfeyer thanks for the reminder! Unfortunately this needs more than just a rebase (see https://github.com/neos/flow-development-collection/pull/130#issuecomment-157188662) but I'll try to tackle it before the release!!

bwaidelich commented:
FYI: 03e3d52 is just a rebase to current master, I still need to adjust the behavior as discussed. stay tuned

kitsunet commented:
Is this ready?

bwaidelich commented:
> Is this ready?

No, it's still missing adjusted documentation. Unfortunately I won't be able to really work on it this week. But anyone feel free to test and/or provide some documentation!

kitsunet commented:
Any news? Would love to get this in for the LTS.

bwaidelich commented:
The examples in the documentation have been adjusted to the new syntax now!

bwaidelich commented:
See  https://github.com/neos/neos-development-collection/pull/648 & https://github.com/Flowpack/Flowpack.Neos.FrontendLogin/pull/12 for (optional) adjustments to related Neos packages

kitsunet commented:
YAY, will review later.

albe commented:
Looks good from reading, will try it out later - don't have a project with patterns around at the moment.

kitsunet commented:
Ok, will merge on Friday when no one objects. Works perfectly in a project with multiple packages each configuring filters. migration applied just fine.

albe commented:
Looks good from testing so far. Only tiny thing I found was that the migration will give two outputs when a pattern can not be converted:
`Adjusted configuration syntax of the "%s" firewall filter...`
and then also
`Could not automatically convert the syntax of the custom firewall filter "%s"...`

that makes it a bit unclear if any change was actually applied or not, but by no means blocking this. I'd say good to merge.
